lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,743 @@
+        super->error.cb = on_subreq_error_callback;
+        super->error.data = subreq;
+    }
+
+    /* headers */
+    super->headers = (h2o_headers_t){NULL};
+    if (h2o_mruby_iterate_headers(ctx->shared, dupenv, handle_request_header, subreq) != 0) {
+        goto Failed;
+    }
+
+    /* entity */
+    prepare_subreq_entity(super, ctx, rack_input);
+    if (mrb->exc != NULL)
+        goto Failed;
+
+    /* env */
+    mrb_value other_keys = mrb_hash_keys(mrb, dupenv);
+    for (i = 0; i != RARRAY_LEN(other_keys); ++i) {
+        mrb_value key = h2o_mruby_to_str(mrb, mrb_ary_entry(other_keys, i));
+        if (memcmp(RSTRING_PTR(key), "HTTP_", 5) == 0)
+            continue;
+        mrb_value val = h2o_mruby_to_str(mrb, mrb_hash_get(mrb, dupenv, key));
+        h2o_vector_reserve(&super->pool, &super->env, super->env.size + 2);
+        super->env.entries[super->env.size] = h2o_strdup(&super->pool, RSTRING_PTR(key), RSTRING_LEN(key));
+        super->env.entries[super->env.size + 1] = h2o_strdup(&super->pool, RSTRING_PTR(val), RSTRING_LEN(val));
+        super->env.size += 2;
+    }
+
+    return subreq;
+
+Failed:
+    assert(mrb->exc != NULL);
+    mrb_gc_arena_restore(mrb, gc_arena);
+    return NULL;
+}
+
+static void on_postfilter_setup_stream(h2o_req_filter_t *self, h2o_req_t *req, h2o_ostream_t **slot)
+{
+    struct st_mruby_subreq_t *subreq = (void *)req;
+    mrb_state *mrb = subreq->ctx->shared->mrb;
+
+    assert(mrb_nil_p(subreq->ref));
+    assert(! mrb_nil_p(subreq->receiver));
+
+    /* at first call, mruby is blocking by H2O.next.call */
+
+    int gc_arena = mrb_gc_arena_save(mrb);
+
+    mrb_value resp = build_app_response(subreq);
+    subreq->ref = mrb_ary_entry(resp, 2);
+
+    /* detach receiver */
+    mrb_value receiver = subreq->receiver;
+    mrb_gc_unregister(mrb, receiver);
+    mrb_gc_protect(mrb, receiver);
+    subreq->receiver = mrb_nil_value();
+    h2o_mruby_run_fiber(subreq->ctx, receiver, resp, NULL);
+
+    mrb_gc_arena_restore(mrb, gc_arena);
+
+
+    h2o_setup_next_req_filter(self, req, slot);
+}
+
+static mrb_value middleware_call_callback(h2o_mruby_context_t *ctx, mrb_value input, mrb_value receiver, mrb_value args, int *run_again)
+{
+    mrb_state *mrb = ctx->shared->mrb;
+
+    mrb_value env = mrb_ary_entry(args, 0);
+    if (! mrb_hash_p(env)) {
+        *run_again = 1;
+        return mrb_exc_new_str_lit(mrb, E_ARGUMENT_ERROR, "env must be a hash");
+    }
+
+    mrb_value reprocess = mrb_ary_entry(args, 1);
+
+    /* create subreq */
+    struct st_mruby_subreq_t *subreq = create_subreq(ctx, env);
+    if (mrb->exc != NULL) {
+        *run_again = 1;
+        return mrb_obj_value(mrb->exc);
+    }
+    subreq->receiver = receiver;
+    mrb_gc_register(mrb, receiver);
+
+    h2o_req_t *super = &subreq->super;
+
+    /* create postfilter and final ostream for subreq */
+    h2o_req_filter_t *postfilter = (void *)h2o_add_req_filter(super, sizeof(*postfilter), 1);
```

Comment:
Do we still need to have postfilter as a new feature now that you have implemented your own sub-request? http1.c and http2/stream.c do not use postfilter and it might be the case that we can do the same for mruby middleware.